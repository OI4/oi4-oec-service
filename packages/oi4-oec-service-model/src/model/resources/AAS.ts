import {OI4Payload} from '../Payload';
import {Resources} from '../Resources';

export class AAS implements OI4Payload {
    readonly AASId: string;
    readonly GlobalAssetId: string;

    constructor(id: string, gId: string) {
        this.AASId = id;
        this.GlobalAssetId = gId;
    }

    resourceType(): Resources {
        return Resources.AAS;
    }

    static clone(source: AAS): AAS {
        return new AAS(source.AASId, source.GlobalAssetId);
    }

}
