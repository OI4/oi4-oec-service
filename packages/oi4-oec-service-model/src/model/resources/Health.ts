import {OI4Payload} from '../Payload';
import {EDeviceHealth} from '../EContainer';
import {Resources} from '../Resources';

export class Health implements OI4Payload {
    readonly Health: EDeviceHealth;
    readonly HealthScore: number; // UInt16 (from 0 to 100%)

    constructor(health: EDeviceHealth, healthScore: number) {
        this.Health = health;
        this.HealthScore = healthScore;
    }

    resourceType(): Resources {
        return Resources.HEALTH;
    }

    static clone(source: Health): Health {
        return new Health(source.Health, source.HealthScore);
    }
}
