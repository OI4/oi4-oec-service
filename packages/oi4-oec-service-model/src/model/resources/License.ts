import {OI4Payload} from '../Payload';
import {Resources} from '../Resources';

export class License implements OI4Payload {
    readonly LicenseId: string;
    readonly Components: IComponentObject[];

    constructor(licenseId: string, components: IComponentObject[]) {
        this.LicenseId = licenseId;
        this.Components = components;
    }

    resourceType(): Resources {
        return Resources.LICENSE;
    }

    static clone(source: License): License {
        return new License(source.LicenseId, source.Components);
    }

}

export interface IComponentObject {
    Component: string;
    LicAuthors: string[];
    LicAddText: string;
}
