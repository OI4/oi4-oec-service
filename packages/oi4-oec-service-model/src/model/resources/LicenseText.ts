import {OI4Payload} from '../Payload';
import {Resources} from '../Resources';

export class LicenseText implements OI4Payload {
    readonly LicenseText: string;

    constructor(licenseText: string) {
        this.LicenseText = licenseText;
    }

    resourceType(): Resources {
        return Resources.LICENSE_TEXT;
    }

    static clone(source: LicenseText): LicenseText {
        return new LicenseText(source.LicenseText);
    }
}
