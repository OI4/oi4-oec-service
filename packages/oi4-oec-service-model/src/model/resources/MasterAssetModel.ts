import {OI4Payload} from '../Payload';
import {IMasterAssetModel, IOPCUALocalizedText} from '../../opcua/model/IOPCUA';
import {Oi4Identifier} from '../Oi4Identifier';
import {ServiceTypes, getServiceType} from '../ServiceTypes';
import {Resources} from '../Resources';

export class MasterAssetModel implements OI4Payload, IMasterAssetModel {
    Description: IOPCUALocalizedText;
    DeviceClass: string;
    DeviceManual: string;
    DeviceRevision: string;
    HardwareRevision: string;
    Manufacturer: IOPCUALocalizedText;
    ManufacturerUri: string;
    Model: IOPCUALocalizedText;
    ProductCode: string;
    ProductInstanceUri: string;
    RevisionCounter: number;
    SerialNumber: string;
    SoftwareRevision: string;

    resourceType(): Resources {
        return Resources.MAM;
    }

    getOI4Id(): Oi4Identifier {
        return new Oi4Identifier(this.ManufacturerUri, this.Model.Text, this.ProductCode, this.SerialNumber);
    }

    getServiceType(): ServiceTypes {
        const serviceType = this.DeviceClass.startsWith('Oi4.') ? this.DeviceClass.substring(4) : this.DeviceClass;
        return getServiceType(serviceType);
    }

    static clone(source: IMasterAssetModel): MasterAssetModel {
        const copy = new MasterAssetModel();
        Object.assign(copy, source);
        return copy;
    }
}
