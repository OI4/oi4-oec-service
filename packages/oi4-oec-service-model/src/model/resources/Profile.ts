import {OI4Payload} from '../Payload';
import {Resources} from '../Resources';

export class Profile implements OI4Payload {
    readonly Resources: Resources[];

    constructor(resources: Resources[]) {
        this.Resources = Object.assign([], resources);
    }

    resourceType(): Resources {
        return Resources.PROFILE;
    }

    static clone(source: Profile): Profile {
        return new Profile(source.Resources);
    }
}
