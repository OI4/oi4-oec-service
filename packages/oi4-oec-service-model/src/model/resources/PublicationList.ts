import {OI4Payload} from '../Payload';
import {Oi4Identifier} from '../Oi4Identifier';
import {Resources} from '../Resources';

export class PublicationList implements OI4Payload {
    Resource: Resources;
    Source: Oi4Identifier;
    Filter?: string;
    DataSetWriterId: number; // Actually OI4-Identifier: TODO: Validator
    Mode: PublicationListMode;
    Interval?: number; // UINT32
    Precisions?: number; // REAL
    Config?: PublicationListConfig;

    resourceType(): Resources {
        return Resources.PUBLICATION_LIST;
    }

    static clone(source: PublicationList): PublicationList {
        const copy = new PublicationList();
        Object.assign(copy, source);
        return copy;
    }
}

export enum PublicationListMode {
    OFF_0 = 'OFF_0',
    ON_REQUEST_1 = 'ON_REQUEST_1',
    APPLICATION_2 = 'APPLICATION_2',
    SOURCE_3 = 'SOURCE_3',
    FILTER_4 = 'FILTER_4',
    APPLICATION_SOURCE_5 = 'APPLICATION_SOURCE_5',
    APPLICATION_FILTER_6 = 'APPLICATION_FILTER_6',
    SOURCE_FILTER_7 = 'SOURCE_FILTER_7',
    APPLICATION_SOURCE_FILTER_8 = 'APPLICATION_SOURCE_FILTER_8',
}

export enum PublicationListConfig {
    NONE_0 = 'NONE_0',
    MODE_1 = 'MODE_1',
    INTERVAL_2 = 'INTERVAL_2',
    MODE_AND_INTERVAL_3 = 'MODE_AND_INTERVAL_3',
}
