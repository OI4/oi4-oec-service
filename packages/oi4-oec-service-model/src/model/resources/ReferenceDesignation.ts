import {OI4Payload} from '../Payload';
import {Resources} from '../Resources';
import {Oi4Identifier} from '../Oi4Identifier';

export class ReferenceDesignation implements OI4Payload {

    readonly Function: ReferenceDesignationObject;
    readonly Product: ReferenceDesignationObject;
    readonly Location: ReferenceDesignationObject;

    constructor(functionObject: ReferenceDesignationObject, productObject: ReferenceDesignationObject, locationObject: ReferenceDesignationObject) {
        this.Function = functionObject;
        this.Product = productObject;
        this.Location = locationObject;
    }

    resourceType(): Resources {
        return Resources.REFERENCE_DESIGNATION;
    }

    static clone(source: ReferenceDesignation): ReferenceDesignation {
        return new ReferenceDesignation(source.Function, source.Product, source.Location);
    }
}

export interface ReferenceDesignationObject {
    Value: string;
    Local: string;
    Parent: ReferenceDesignationParent;
}

export interface ReferenceDesignationParent {
    Value: string;
    Local: string;
    Oi4Identifier: Oi4Identifier;
}
