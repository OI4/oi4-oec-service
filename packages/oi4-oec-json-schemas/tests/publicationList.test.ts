import { matchersWithOptions } from 'jest-json-schema'
import schema from '../src/schemas/PublicationList.schema.json'
import uint16 from '../src/schemas/dataTypes/uint16.schema.json'
import uint32 from '../src/schemas/dataTypes/uint32.schema.json'
import resources from '../src/schemas/constants/resources.schema.json'
import oi4Identifier from '../src/schemas/Oi4Identifier.schema.json'

import validObjs from './__fixtures__/publicationList_valid.json'
import invalidObjs from './__fixtures__/publicationList_invalid.json'

expect.extend(
    matchersWithOptions({
      // Loading in a schema which is composed only of definitions,
      // which means specific test schemas need to be created.
      // This is good for testing specific conditions for definition schemas.
      schemas: [uint16, uint32, resources, oi4Identifier],
      verbose: true,
    })
  )
describe('publicationList schema', () => {
  it('validate schema', () => {
    expect(schema).toBeValidSchema()
  })

  it.each(validObjs as [])(
    '(%#) match valid publicationList object to schema -> %s',
    (_name: string, obj) => {
      expect(obj).toMatchSchema(schema)
    }
  )

  it.each(invalidObjs as [])(
    '(%#) match fails for invalid publicationList -> %s',
    (name: string, obj) => {
      expect(obj).not.toMatchSchema(schema)
      expect(() => {
        expect(obj).toMatchSchema(schema)
      }).toThrowErrorMatchingSnapshot(name)
    }
  )
})
