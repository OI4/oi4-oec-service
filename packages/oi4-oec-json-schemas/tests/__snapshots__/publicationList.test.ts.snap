// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`publicationList schema (0) match fails for invalid publicationList -> empty: empty 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  must have required property 'Resource'[39m
[31m    Path:     publicationList.schema.json#/required[39m
[31m[39m[31m  must have required property 'DataSetWriterId'[39m
[31m    Path:     publicationList.schema.json#/required[39m
[31m[39m[31m  must have required property 'Source'[39m
[31m    Path:     publicationList.schema.json#/required[39m
[31m[39m"
`;

exports[`publicationList schema (1) match fails for invalid publicationList -> publicationList with unknown resource: publicationList with unknown resource 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  /Resource must be equal to one of the allowed values[39m
[31m    Expected: [\\"MAM\\",\\"Health\\",\\"Config\\",\\"License\\",\\"LicenseText\\",\\"RtLicense\\",\\"Data\\",\\"Metadata\\",\\"Event\\",\\"Profile\\",\\"PublicationList\\",\\"SubscriptionList\\",\\"ReferenceDesignation\\",\\"Interfaces\\"][39m
[31m    Received: [33m<string>[39m[31m unknown[39m
[31m    Path:     publicationList.schema.jsonconstants/resources.schema.json/enum[39m
[31m[39m"
`;

exports[`publicationList schema (2) match fails for invalid publicationList -> publicationList with missing Resource: publicationList with missing Resource 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  must have required property 'Resource'[39m
[31m    Path:     publicationList.schema.json#/required[39m
[31m[39m"
`;

exports[`publicationList schema (3) match fails for invalid publicationList -> publicationList with missing DataSetWriterId: publicationList with missing DataSetWriterId 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  must have required property 'DataSetWriterId'[39m
[31m    Path:     publicationList.schema.json#/required[39m
[31m[39m"
`;

exports[`publicationList schema (4) match fails for invalid publicationList -> publicationList with missing Source: publicationList with missing Source 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  must have required property 'Source'[39m
[31m    Path:     publicationList.schema.json#/required[39m
[31m[39m"
`;

exports[`publicationList schema (5) match fails for invalid publicationList -> publicationList with invalid Source: publicationList with invalid Source 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toMatchSchema([22m[32mschema[39m[2m)[22m

received
[31m  /Source must match pattern \\"^(([a-z0-9-]+\\\\.)*([a-z0-9-]*)(?:/[^/\`\\\\^[39m
[31m]+){3})$\\"[39m
[31m    Received: [33m<string>[39m[31m provider.com[39m
[31m    Path:     publicationList.schema.jsonoi4Identifier.schema.json/pattern[39m
[31m[39m"
`;
